sonar-ws/src/main/java/org/sonarqube/ws/client/HttpConnector.java

```diff
@@ -17,73 +17,192 @@
+import javax.annotation.CheckForNull;
 import javax.annotation.Nullable;
 
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.base.Strings.nullToEmpty;
+import static java.lang.String.format;
+
 public class HttpConnector implements WsConnector {
 
-  public static final int DEFAULT_CONNECT_TIMEOUT_MILLISECONDS = 30000;
-  public static final int DEFAULT_READ_TIMEOUT_MILLISECONDS = 60000;
+  public static final int DEFAULT_CONNECT_TIMEOUT_MILLISECONDS = 30_000;
+  public static final int DEFAULT_READ_TIMEOUT_MILLISECONDS = 60_000;
 
   /**
-   * Visibility relaxed for unit tests
+   * Base URL with trailing slash, for instance "https://localhost/sonarqube/".
+   * It is required for further usage of {@link HttpUrl#resolve(String)}.
    */
-  final HttpRequestFactory requestFactory;
+  private final HttpUrl baseUrl;
+  private final String userAgent;
+  private final String credentials;
+  private final String proxyCredentials;
+  private final OkHttpClient okHttpClient = new OkHttpClient();
 
   private HttpConnector(Builder builder) {
-    this.requestFactory = new HttpRequestFactory(builder.url)
-      .setLogin(builder.login)
-      .setPassword(builder.password)
-      .setProxyHost(builder.proxyHost)
-      .setProxyPort(builder.proxyPort)
-      .setProxyLogin(builder.proxyLogin)
-      .setProxyPassword(builder.proxyPassword)
-      .setConnectTimeoutInMilliseconds(builder.connectTimeoutMs)
-      .setReadTimeoutInMilliseconds(builder.readTimeoutMs);
+    this.baseUrl = HttpUrl.parse(builder.url.endsWith("/") ? builder.url : format("%s/", builder.url));
+    this.userAgent = builder.userAgent;
+
+    if (isNullOrEmpty(builder.login)) {
+      // no login nor access token
+      this.credentials = null;
+    } else {
+      // password is null when login represents an access token. In this case
+      // the Basic credentials consider an empty password.
+      this.credentials = Credentials.basic(builder.login, nullToEmpty(builder.password));
+    }
+
+    if (builder.proxy != null) {
+      this.okHttpClient.setProxy(builder.proxy);
+    }
+    // proxy credentials can be used on system-wide proxies, so even if builder.proxy is null
+    if (isNullOrEmpty(builder.proxyLogin)) {
+      this.proxyCredentials = null;
+    } else {
+      this.proxyCredentials = Credentials.basic(builder.proxyLogin, nullToEmpty(builder.proxyPassword));
+    }
+
+    this.okHttpClient.setConnectTimeout(builder.connectTimeoutMs, TimeUnit.MILLISECONDS);
+    this.okHttpClient.setReadTimeout(builder.readTimeoutMs, TimeUnit.MILLISECONDS);
+    this.okHttpClient.interceptors().addAll(builder.interceptors);
   }
 
   @Override
-  public String execute(WsRequest wsRequest) {
-    return requestFactory.execute(wsRequest);
+  public String baseUrl() {
+    return baseUrl.url().toExternalForm();
+  }
+
+  public OkHttpClient okHttpClient() {
```

Comment:
as said previously, I wouldn't expose the internal http client, specific to a library